pkg/controller/agent/nad/controller.go

```diff
@@ -40,14 +42,25 @@ func (h Handler) OnChange(key string, nad *nadv1.NetworkAttachmentDefinition) (*
 	if nad == nil || nad.DeletionTimestamp != nil {
 		return nil, nil
 	}
-	if !utils.IsVlanNAD(nad) {
-		return nad, nil
-	}
 
-	klog.Infof("nad configuration %s has been changed: %s", nad.Name, nad.Spec.Config)
+	klog.Infof("nad configuration %s/%s has been changed: %s", nad.Namespace, nad.Name, nad.Spec.Config)
+
+	// nadCopy could be nil
+	nadCopy, err := h.removeOutdatedLocalArea(nad)
+	if err != nil {
+		return nil, fmt.Errorf("remove outdated local area for nad %s/%s failed, error: %w", nad.Namespace, nad.Name, err)
+	}
 
+	// NadCaopy could be nil and nad is readonly here
 	if err := h.addLocalArea(nad); err != nil {
-		return nil, err
+		return nil, fmt.Errorf("add local area for nad %s/%s failed, error: %w", nad.Namespace, nad.Name, err)
+	}
+
+	// update nad if needed
+	if nadCopy != nil {
```

Comment:
Hint:
The general practice is to return the updated object when it is successful, let other OnChange be invoked with the updated object instead of old one.